R/opts.R

```diff
@@ -442,6 +461,98 @@ opt_table_outline <- function(data,
   # Perform input object validation
   stop_if_not_gt(data = data)
 
-  values_vec <- rep(c(style, width, color), 4)
-  tab_options_multi(data, create_option_value_list(table_borders_vec(), values_vec))
+  if (style == "none") {
+    width <- NULL
+    color <- NULL
+  }
+
+  params <-
+    c(style = !is.null(style), width = !is.null(width), color = !is.null(color)) %>%
+    which() %>%
+    names()
+
+  pattern <- paste0("^table\\.border.*?\\.(", paste(params, collapse = "|"), ")")
+
+  # Get vector of `tab_options()` arg names for the table border styles
+  options_vec <- get_tab_options_arg_vec(pattern = pattern)
+
+  if (is.null(width)) width <- NA_character_
+  if (is.null(color)) color <- NA_character_
+
+  values_vec <-
+    dplyr::case_when(
+      grepl("^table\\.border.*?\\.style$", options_vec) ~ as.character(style),
+      grepl("^table\\.border.*?\\.width$", options_vec) ~ as.character(width),
+      grepl("^table\\.border.*?\\.color$", options_vec) ~ as.character(color)
+    )
+
+  option_value_list <- create_option_value_list(options_vec, values_vec)
+  tab_options_multi(data, option_value_list)
+}
+
+# Create an option-value list with a vector of arg names from the
+# `tab_options()` function and either one value or n-length values
+# corresponding to those options
+create_option_value_list <- function(tab_options_args, values) {
+
+  # Validate the length of the `values` vector
+  if (length(values) == 1) {
+    values <- rep(values, length(tab_options_args))
+  } else if (length(values) != length(tab_options_args)) {
+    stop("The length of the `values` vector must be 1 or the length of `tab_options_args`")
+  }
+
+  # Validate the elements of the `tab_options_args` vector
+  validate_tab_options_args(tab_options_args)
+
+  stats::setNames(object = values, tab_options_args) %>% as.list()
+}
+
+create_default_option_value_list <- function(tab_options_args) {
+
+  # Validate the elements of the `tab_options_args` vector
+  validate_tab_options_args(tab_options_args)
+
+  lapply(stats::setNames(, tab_options_args), function(x) {
+    dt_options_get_default_value(tidy_gsub(x, ".", "_", fixed = TRUE))
+  })
+}
+
+dt_options_get_default_value <- function(option) {
+
+  # Validate the provided `option` value
+  if (length(option) != 1) {
+    stop("A character vector of length one must be provided")
+  }
+  if (!(option %in% dt_options_tbl$parameter)) {
+    stop("The `option` provided is invalid")
+  }
+
+  dt_options_tbl$value[[which(dt_options_tbl$parameter == option)]]
+}
+
+# Validate any vector of `tab_options()` argument names
+validate_tab_options_args <- function(tab_options_args) {
+
+  tab_options_arg_names <-
+    formals(tab_options) %>% names() %>% base::setdiff("data")
```

Comment:
`tab_options_arg_names` could be made into a top-level variable. To ensure `tab_options` is defined by the time we get here, use `#' @include tab-options.R`.

Reply:
All done. This is great!